authorization/role/service/role_management_service.go

```diff
@@ -50,84 +50,92 @@ func (r *RoleManagementServiceImpl) ListAvailableRolesByResourceType(ctx context
 		return err
 	}
 
 	// check if the current user token belongs to a user who has the necessary privileges
 	// for assigning roles to other users.
-
 	permissionService := permservice.NewPermissionService(r.repo)
 	hasScope, err := permissionService.HasScope(ctx, assignedBy, resourceID, authorization.ScopeForManagingRolesInResourceType(rt.Name))
 	if err != nil {
 		log.Error(ctx, map[string]interface{}{
 			"resource_id": resourceID,
 			"identity_id": assignedBy,
-			"role":        roleName,
-		}, "error determining if user has manage scope")
+		}, "error determining if user may manage roles for resource")
 		return err
 	}
 	if !hasScope {
 		log.Error(ctx, map[string]interface{}{
 			"resource_id": resourceID,
 			"identity_id": assignedBy,
-			"role":        roleName,
 		}, "user not authorizied to assign roles")
 		return errors.NewForbiddenError("user is not authorized to assign roles")
 	}
 
-	// In batch assignment of roles all selected users must have previously been assigned
+	// Valid all the roles and user identity IDs, and ensure each user has been previously assigned
 	// privileges for the resource, otherwise the invitation workflow should be used instead
-	for _, identityIDAsUUID := range identitiesToBeAssigned {
-		assignedRoles, err := r.repo.IdentityRoleRepository().FindIdentityRolesByIdentityAndResource(ctx, resourceID, identityIDAsUUID)
+	assignments := make(map[uuid.UUID][]uuid.UUID)
+
+	for k, v := range roleAssignments {
+		roleRef, err := r.repo.RoleRepository().Lookup(ctx, k, rt.ResourceType.Name)
 		if err != nil {
-			log.Error(ctx, map[string]interface{}{
-				"resource_id": resourceID,
-				"identity_id": assignedBy,
-				"role":        roleName,
-			}, "error looking up existing assignments")
 			return err
 		}
-		if len(assignedRoles) == 0 {
-			log.Error(ctx, map[string]interface{}{
-				"resource_id": resourceID,
-				"identity_id": identityIDAsUUID,
-				"role":        roleName,
-			}, "identity not part of a resource cannot be assigned a role")
-			return errors.NewBadParameterErrorFromString("identityID", identityIDAsUUID, fmt.Sprintf("cannot update roles for an identity %s without an existing role", identityIDAsUUID))
+
+		for _, identityIDAsUUID := range v {
+			assignedRoles, err := r.repo.IdentityRoleRepository().FindIdentityRolesByIdentityAndResource(ctx, resourceID, identityIDAsUUID)
+			if err != nil {
+				log.Error(ctx, map[string]interface{}{
+					"resource_id": resourceID,
+					"identity_id": assignedBy,
+				}, "error looking up existing assignments")
+				return err
+			}
+			if len(assignedRoles) == 0 {
+				log.Error(ctx, map[string]interface{}{
+					"resource_id": resourceID,
+					"identity_id": identityIDAsUUID,
+				}, "identity not previously assigned a resource role cannot be assigned another role")
+				return errors.NewBadParameterErrorFromString("identityID", identityIDAsUUID, fmt.Sprintf("cannot update roles for an identity %s without an existing role", identityIDAsUUID))
```

Comment:
Just realized that we can't actually use this service for initial role assignment (during resource creation for example) when the user doesn't have any role in the resource yet.

We need to add a new method to the service for that or add a param to Assign()

Reply:
I will add a new method for that in https://github.com/fabric8-services/fabric8-auth/pull/498